fix: reset object before reading in new values

pihme opened the pull request:
## Description

This is one of the things where you spend hours debugging something very peculiar, and then the fix is a simple line of code.
However, since this line of code is at a prominent spot, I thought I make it a PR of its own. Maybe there are more side effects, or maybe there are other places which need to get changed as well.

Without the reset there are problems if the schema changes. Undeclared properties will accumulate,
because each read will add the same undeclared property to the list.
![2561084e86c7466ab6f6b35b88b4f2cd](https://user-images.githubusercontent.com/14032870/119957365-4ea55f80-bfa2-11eb-8e0c-86dcfd4940cc.png)
Worse yet, on each write the accumulated list of undeclared properties will be written, so this can grow quite fast.

Also properties which are not present in the data stream will keep the value they currently have,
instead of resetting to their default value.

<!-- Please explain the changes you made here. -->

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda-cloud/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [X] I've reviewed my own code
* [X] I've written a clear changelist description
* [X] I've narrowly scoped my changes
* [X] I've separated structural from behavioural changes
-->

## Definition of Done

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [X] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

saig0 reviewed (approved):
Good catch :+1: 

If possible and feasible then please add a test case to verify the fix.

pihme commented:
@saig0 Please have another look. I added some tests, but I am not very happy with them. I think I won't understand them when I look at them in a couple of months. So any suggestions are very welcome. Also I changed an existing tests that failed. Not sure if the detail I changed was important for some other places/assumptions about the behavior of the class.

saig0 commented on msgpack-value/src/test/java/io/camunda/zeebe/msgpack/UnpackedObjectTest.java:
```diff
@@ -0,0 +1,91 @@
+      spyUnpackedObject.wrap(buffer);
+
+      // then
+      final var orderOfInvocations = Mockito.inOrder(spyUnpackedObject);
+      orderOfInvocations.verify(spyUnpackedObject).reset();
+      orderOfInvocations.verify(spyUnpackedObject).read(Mockito.any());
```
This test case is not robust because it depends on implementation details. I don't see much value in it.

saig0 commented on msgpack-value/src/test/java/io/camunda/zeebe/msgpack/UnpackedObjectTest.java:
```diff
@@ -0,0 +1,91 @@
+          .describedAs("value of added property after reading")
+          .isFalse();
+    }
+  }
```
I like this test case :+1: 

We could add another test case for the behavior that you described in the pull request. For example:

```
    @Test
    public void shouldNotAccumulateSizeWithUndeclaredProperties() {
      // given
      newSchemaObject.wrap(bufferSerializedWithOldSchema);
      final int length = newSchemaObject.getLength();

      final var buffer = new UnsafeBuffer(ByteBuffer.allocate(100));
      newSchemaObject.write(buffer, 0);

      // when
      newSchemaObject.wrap(buffer);

      // then
      assertThat(newSchemaObject.getLength()).isEqualTo(length);
    }
```

saig0 reviewed (approved):
I'm not sure about the `GreyBox` test but I like the `SchemaEvolution` test.

Please have a look at my suggestion for another test case.
Feel free to merge the PR :rocket:

pihme commented:
bors r+
